### PR TITLE
CEDS 2159 - Display Movement Query Response to Loaders

### PR DESCRIPTION
**Pull request by SteveSugden**
Functionally complete but may require addressing the following
- what to display if feature switch disabled when accessing new end-points
- changes to customs-declare-exports-movements for query time-out

### CODE REVIEW

**Comment by SteveSugden** [comment]
Ready for review now.  I've integrated with the latest changes for query time-out made to the back-end service and the internal movements front-end

File: app/controllers/ChoiceController.scala [hunk 1]
```diff
@@ -35,14 +35,14 @@ class ChoiceController @Inject()(authenticate: AuthAction, cache: CacheRepositor
...
-        case Some(choice) => Ok(choicePage(Choice.form().fill(choice)))
-        case None         => Ok(choicePage(Choice.form()))
-      }
+  def displayChoiceForm: Action[AnyContent] = authenticate.async { implicit request =>
+    cache.findByEori(request.eori).map {
+      case Some(cache) =>
+        cache.answers
+          .map(answers => Ok(choicePage(Choice.form().fill(Choice(answers.`type`)))))
+          .getOrElse(Ok(choicePage(Choice.form())))
+      case None => Ok(choicePage(Choice.form()))
```
Comment (MaciejRewera) [inline app/controllers/ChoiceController.scala]: Isn't the second case redundant? There is a repetition, so I thought maybe it could be unified?

File: app/controllers/actions/IleQueryAction.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import play.api.mvc._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class IleQueryAction @Inject()(appConfig: AppConfig)(implicit val exc: ExecutionContext)
+    extends ActionFunction[AuthenticatedRequest, AuthenticatedRequest] {
+  override def invokeBlock[A](request: AuthenticatedRequest[A], block: AuthenticatedRequest[A] => Future[Result]): Future[Result] =
+    if (appConfig.ileQueryEnabled) block(request) else throw FeatureDisabledException
+  override protected def executionContext: ExecutionContext = exc
+}
```
Comment (MaciejRewera) [inline app/controllers/actions/IleQueryAction.scala]: Nit: Could you make empty lines between methods? :)

File: app/controllers/ChoiceController.scala (hunk 1 above)
Comment (SteveSugden) [reply app/controllers/ChoiceController.scala]: Copied but I'll take a look.

File: app/handlers/ErrorHandler.scala [hunk 3]
```diff
@@ -60,4 +61,7 @@ class ErrorHandler @Inject()(appConfig: AppConfig, override val messagesApi: Mes
         message = Messages("global.error.message")
       )
     )
+
+  def standardErrorTemplate()(implicit request: Request[_]): Html =
+    errorTemplate("global.error.title", "global.error.heading", "global.error.message")
```
Comment (MaciejRewera) [inline app/handlers/ErrorHandler.scala]: Is this being used anywhere?

Comment (SteveSugden) [reply app/handlers/ErrorHandler.scala]: Twice in the IleQueryController

File: app/controllers/actions/IleQueryAction.scala (hunk 2 above)
Comment (SteveSugden) [reply app/controllers/actions/IleQueryAction.scala]: Doing

File: app/controllers/ChoiceController.scala (hunk 1 above)
Comment (SteveSugden) [reply app/controllers/ChoiceController.scala]: Done.  By the power of flatmapping - ```  def displayChoiceForm: Action[AnyContent] = authenticate.async { implicit request =>
    cache.findByEori(request.eori).map(_.flatMap(_.answers)).map {
      case Some(answers) => Ok(choicePage(Choice.form().fill(Choice(answers.`type`))))
      case None          => Ok(choicePage(Choice.form()))
    }
  }```

File: app/models/notifications/queries/IleQueryResponseExchangeData.scala [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  case class SuccessfulResponseExchangeData(
+    queriedDucr: Option[DucrInfo] = None,
+    queriedMucr: Option[MucrInfo] = None,
+    parentMucr: Option[MucrInfo] = None,
+    childDucrs: Seq[DucrInfo] = Seq.empty,
+    childMucrs: Seq[MucrInfo] = Seq.empty
+  ) extends IleQueryResponseExchangeData {
+    override val typ = IleQueryResponseExchangeType.SuccessfulResponseExchange
+
+    val sortedChildrenUcrs: Seq[UcrInfo] = (childDucrs ++ childMucrs).sortBy(_.entryStatus.flatMap(_.roe).getOrElse(ROECode.NoneRoe))
```
Comment (MaciejRewera) [inline app/models/notifications/queries/IleQueryResponseExchangeData.scala]: Nit: Is it worth making it `lazy` ?

File: app/models/notifications/queries/MucrInfo.scala [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+import models.notifications.EntryStatus
+import play.api.libs.json.Json
+
+case class MucrInfo(
+  ucr: String,
+  parentMucr: Option[String] = None,
+  entryStatus: Option[EntryStatus] = None,
+  isShut: Option[Boolean] = None,
+  movements: Seq[MovementInfo] = Seq.empty
```
Comment (MaciejRewera) [inline app/models/notifications/queries/MucrInfo.scala]: Actually, doesn't it need to `override` ?

File: app/models/notifications/queries/IleQueryResponseExchangeData.scala (hunk 4 above)
Comment (SteveSugden) [reply app/models/notifications/queries/IleQueryResponseExchangeData.scala]: Again, copied.  But I'll change.

File: app/models/notifications/queries/MucrInfo.scala (hunk 5 above)
Comment (SteveSugden) [reply app/models/notifications/queries/MucrInfo.scala]: Apparently not.

File: app/connectors/CustomsDeclareExportsMovementsConnector.scala [hunk 6]
```diff
@@ -37,6 +37,8 @@ class CustomsDeclareExportsMovementsConnector @Inject()(appConfig: AppConfig, ht
 
   private val logger = Logger(this.getClass)
 
+  private val FailedDependencyStatus: Int = 424
```
Comment (PauloPortugal) [inline app/connectors/CustomsDeclareExportsMovementsConnector.scala]: I would probably use the `Status.FAILED_DEPENDENCY` instead of the _"magic number **424**"_

Comment (SteveSugden) [reply app/connectors/CustomsDeclareExportsMovementsConnector.scala]: Good idea

File: test/unit/models/notifications/IleQueryResponseExchangeDataSpec.scala [hunk 7]
```diff
@@ -43,4 +43,40 @@ class IleQueryResponseExchangeDataSpec extends UnitSpec {
       response.sortedChildrenUcrs.map(_.entryStatus.flatMap(_.roe).get) mustBe expectedStatusOrder
     }
   }
+
+  "SuccessfulResponseExchangeData on queriedUcr" when {
```
Comment (MaciejRewera) [inline test/unit/models/notifications/IleQueryResponseExchangeDataSpec.scala]: Why did you remove one test case?
